passes/cmds/wrapcell.cc

```diff
@@ -114,43 +187,88 @@ struct WrapcellPass : Pass {
 		CellTypes ct;
 		ct.setup();
 
+		bool tracking_unused = has_fmt_field(name_fmt, "%unused");
+
 		for (auto module : d->selected_modules()) {
+			SigPool unused;
+
+			for (auto wire : module->wires())
+			if (wire->has_attribute(ID::unused_bits)) {
+				std::string str = wire->get_string_attribute(ID::unused_bits);
+				for (auto it = str.begin(); it != str.end();) {
+					auto sep = it;
+					for (; sep != str.end() && *sep != ' '; sep++);
+					unused.add(SigBit(wire, std::stoi(std::string(it, sep))));
+					for (it = sep; it != str.end() && *it == ' '; it++);
+				}
+			}
+
 			for (auto cell : module->selected_cells()) {
-				std::optional<std::string> unescaped_name = format(name_fmt, cell->parameters);
+				Module *subm;
+				Cell *subcell;
+
+				if (!ct.cell_known(cell->type))
+					log_error("Non-internal cell type '%s' on cell '%s' in module '%s' unsupported\n",
+							  log_id(cell->type), log_id(cell), log_id(module));
+
+				std::vector<std::pair<IdString, int>> unused_outputs, used_outputs;
+				for (auto conn : cell->connections()) {
+					if (ct.cell_output(cell->type, conn.first))
+					for (int i = 0; i < conn.second.size(); i++) {
+						if (tracking_unused && unused.check(conn.second[i]))
+							unused_outputs.emplace_back(conn.first, i);
+						else
+							used_outputs.emplace_back(conn.first, i);
+					}
+				}
+
+				ContextData context;
+				if (!unused_outputs.empty()) {
+					context.unused_outputs += "_unused";
+					for (auto chunk : collect_chunks(unused_outputs))
+						context.unused_outputs += "_" + RTLIL::unescape_id(chunk.format(cell));
+				}
+
+				std::optional<std::string> unescaped_name = format(name_fmt, cell->parameters, context);
 				if (!unescaped_name)
 					log_error("Formatting error when processing cell '%s' in module '%s'\n",
 							  log_id(cell), log_id(module));
 
 				IdString name = RTLIL::escape_id(unescaped_name.value());
+				if (d->module(name))
+					goto replace_cell;
 
-				if (d->module(name)) {
-					cell->type = name;
-					cell->parameters.clear();
-					continue;
+				subm = d->addModule(name);
+				subcell = subm->addCell("$1", cell->type);
+				for (auto conn : cell->connections()) {
+					if (ct.cell_output(cell->type, conn.first)) {
+						subcell->setPort(conn.first, SigSpec(RTLIL::Sm, conn.second.size()));
```

Comment:
I thought the marker state was only used for temporary testing, but it's load-bearing - when I comment out setPort, the replace calls later in the file fail. Probably worth adding some comments as well. Is it acting as a placeholder as you swap the bits later? Why do `unused_outputs` still get created if they're unused and not labeled `port_output`?

Reply:
> Is it acting as a placeholder as you swap the bits later?

Yes. The fact that marker bits are disallowed in an RTLIL check works to our benefit because if we forget to replace some of the bits we will get caught. I've added a comment

> Why do unused_outputs still get created if they're unused and not labeled port_output?

Output ports must be connected to a wire even if unused, so we need to create those internal wires and connect them to the wrapped cell